common-utils/src/main/java/org/daisy/common/shell/BinaryFinder.java

```diff
@@ -47,5 +70,185 @@
+	 * people may not realize that in general these variables are not
+	 * available in Mac OS applications, we try to make them available
+	 * anyway.
+	 *
+	 * /usr/libexec/path_helper is a utility intended to be used by
+	 * shells for constructing the PATH environment variable. It reads
+	 * from the files /etc/paths and /etc/paths.d/*.
+	 */
+	static Iterable<String> pathFromPathHelper() {
+		return new PathFromPathHelper();
+	}
+	
+	static class PathFromPathHelper implements Iterable<String> {
+		public Iterator<String> iterator() {
+			String pathHelperExec = "/usr/libexec/path_helper";
+			if (new File(pathHelperExec).isFile()) {
+				logger.debug("invoking: " + pathHelperExec + " -s");
+				ProcessBuilder builder = new ProcessBuilder(pathHelperExec, "-s");
+				try {
+					Process process = builder.start();
+					process.waitFor();
+					BufferedReader br = new BufferedReader(new InputStreamReader(process.getInputStream()));
+					try {
+						String line;
+						while ((line = br.readLine()) != null) {
+							logger.debug(line);
+							if (line.startsWith("PATH=\"") && line.endsWith("\"; export PATH;"))
+								return pathFromString(line.substring(6, line.length() - 15), File.pathSeparator).iterator();
+						}
+					} finally {
+						br.close();
+					}
+				} catch(IOException e) {
+					logger.debug("failed: " + e);
+				} catch(InterruptedException e) {
+					logger.debug("failed: " + e);
+				}
+			}
+			return emptyList.iterator();
+		}
+	}
+	
+	/*
+	 * The PATH may be set in shell startup files such as
+	 * /etc/profile, ~/.profile, /etc/bash.bashrc, ~/.bash_profile,
+	 * ~/.bash_login, ~/.bashrc, /etc/csh.login and /etc/zshenv. Some
+	 * of these files invoke /usr/libexec/path_helper but they can
+	 * also set the PATH in other ways.
+	 */
+	static Iterable<String> pathFromShell() {
+		return new PathFromShell();
+	}
+	
+	static class PathFromShell implements Iterable<String> {
+		public Iterator<String> iterator() {
+			String shellExec = System.getenv("SHELL");
+			if (shellExec == null)
+				shellExec = "/bin/sh";
+			if (new File(shellExec).isFile()) {
+				logger.debug("invoking: " + shellExec + " -c \"echo $PATH\"");
+				ProcessBuilder builder = new ProcessBuilder(shellExec, "-c", "echo $PATH");
+				try {
+					Process process = builder.start();
+					process.waitFor();
+					Scanner s = new Scanner(process.getInputStream()).useDelimiter("\\A");
+					try {
+						if (s.hasNext()) {
+							String output = s.next();
+							logger.debug(output);
+							return pathFromString(output, File.pathSeparator).iterator();
+						}
+					} finally {
+						s.close();
+					}
+				} catch(IOException e) {
+					logger.debug("failed: " + e);
+				} catch(InterruptedException e) {
+					logger.debug("failed: " + e);
+				}
+			}
+			return emptyList.iterator();
+		}
+	}
+	
+	private final static Iterable<String> emptyList = emptyList();
+	
+	static <T> Iterable<T> removeDuplicates(final Iterable<T> iterable) {
```

Comment:
Yes I guess so but I like lazy evaluation. The one liner would always evaluate everything (including the `/usr/libexec/path_helper` and `/bin/sh` commands) even if it is not needed. Also I'm not sure the set would preserve the order.